### PR TITLE
Semantic equals implementation

### PR DESCRIPTION
**Pull request by igor-karpukhin**
### All Submissions:

This PR adds generic comparator and normalizer

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes-template.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

File: pkg/api/v1/alert_configurations.go [hunk 1]
```diff
@@ -27,6 +30,15 @@ type AlertConfiguration struct {
 	MetricThreshold *MetricThreshold `json:"metricThreshold,omitempty"`
 }
 
+func (in AlertConfiguration) Key() string {
```
[the remark this answers is not in the document]
Comment (igor-karpukhin) [reply pkg/api/v1/alert_configurations.go:33]: We can remove that, but let's do that later. This change won't affect our current JSONCopy mechanism for AlertConfigurations.

File: pkg/api/v1/alert_configurations.go [hunk 2]
```diff
@@ -208,6 +228,31 @@ type Notification struct {
...
+		in.OpsGenieRegion + "|" +
+		in.OrgName + "|" +
+		in.ServiceKeyRef.Key() + "|" +
+		strconv.FormatBool(admin.GetOrDefault(in.SMSEnabled, false)) + "|" +
+		in.TeamID + "|" +
+		in.TeamName + "|" +
+		in.TypeName + "|" +
+		in.Username + "|" +
+		in.VictorOpsSecretRef.Key() + "|" +
+		"[" + strings.Join(in.Roles, ",") + "]"
```
Comment (josvazg) [inline pkg/api/v1/alert_configurations.go:253]: Why not use a hash code for all the strings instead? does it need to be readable? I to not think will ever be (readable)

Comment (s-urbaniak) [reply pkg/api/v1/alert_configurations.go:253]: hashes are not sortable which is the point of this function

Comment (igor-karpukhin) [reply pkg/api/v1/alert_configurations.go:253]: This is just an example of implementation. It has no impact on the current logic

File: pkg/api/v1/alert_configurations.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/api/v1/alert_configurations.go:33]: sounds good!

**Comment by igor-karpukhin** [comment]
@josvazg @s-urbaniak if your comments were resolved, could you please approve this PR so I could continue working on the Search Indexes implementation?

File: .licenses-gomod.sha256 [hunk 3]
```diff
@@ -1 +1 @@
-100644 d537760138665c24225e4f5725c5bd80775033a6 go.mod
+100644 60d2b0501dda29d0d258c6a0f401c95854606d5c go.mod
```
Comment (josvazg) [inline .licenses-gomod.sha256:1]: no changes to licenses? I guess not or tests would have failed.

File: pkg/api/v1/alert_configurations.go (hunk 2 above)
Comment (s-urbaniak) [reply pkg/api/v1/alert_configurations.go:253]: yes, we need sortable keys for normalization

**Review by s-urbaniak — APPROVED** [review]

File: internal/cmp/normalize.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	return false, -1
+}
+
+func PermuteOrder(data any, r *rand.Rand) {
+	traverse(data, func(slice reflect.Value) {
+		r.Shuffle(slice.Len(), func(i, j int) {
+			reflect.Swapper(slice.Interface())(i, j)
+		})
+	})
+}
```
Comment (josvazg) [inline internal/cmp/normalize.go:66]: This one is only needed for test code right?

File: pkg/api/v1/alert_configurations.go (hunk 2 above)
Comment (s-urbaniak) [reply pkg/api/v1/alert_configurations.go:253]: oh, you indeed have a point about being stable. as long as that invariant is maintained, then yes, stable hashes are good enough

**Review by josvazg — APPROVED** [review]
LGTM

I think we could move a function away to test code, but rest looks good

File: internal/cmp/normalize.go (hunk 4 above)
Comment (s-urbaniak) [reply internal/cmp/normalize.go:66]: potentially but similar as hashmaps in golang it could be a good practice to shuffle nested slices by default

**Review by helderjs — APPROVED** [review]
LGTM, but we still need cover it with more tests right?

File: internal/cmp/normalize.go (hunk 4 above)
Comment (igor-karpukhin) [reply internal/cmp/normalize.go:66]: Yes, this one is only used for testing. I'll try to clear that in a separate PR.

**Comment by igor-karpukhin** [comment]
> LGTM, but we still need cover it with more tests right?

@helderjs You mean to test functions that perform sorting and comparison?

**Pull request merged**